### PR TITLE
Enhancing curation : get_potential_auto_merge()

### PR DESCRIPTION
**Pull request by yger**
This PR improve `get_potential_auto_merge()` with:
  * several metrics for templates distances
  * handling sparsity of templates
  * use computed units locations from analyzer

@DradeAW  : can you check this, if you use this in Lussac. Normally the behavior is the same as before.

### CODE REVIEW

File: src/spikeinterface/curation/auto_merge.py [hunk 1]
```diff
@@ -194,10 +204,18 @@ def get_potential_auto_merge(
             templates_ext is not None
         ), "auto_merge with template_similarity requires a SortingAnalyzer with extension templates"
 
-        templates = templates_ext.get_templates(operator="average")
+        templates = templates_ext.get_data(outputs="Templates")
+        templates = templates.to_sparse(sorting_analyzer.sparsity)
```
Comment (samuelgarcia) [inline src/spikeinterface/curation/auto_merge.py:208]:
```suggestion
        if sorting_analyzer.sparsity is not None:
            templates = templates.to_sparse(sorting_analyzer.sparsity)
```

File: src/spikeinterface/curation/auto_merge.py [hunk 2]
```diff
@@ -403,30 +424,53 @@ def compute_templates_diff(sorting, templates, num_channels=5, num_shift=5, pair
...
     if pair_mask is None:
         pair_mask = np.ones((n, n), dtype="bool")
 
+    if isinstance(templates, Templates):
+        adaptative_masks = (num_channels == None) and (templates.sparsity is not None)
+        if templates.sparsity is not None:
+            sparsity = templates.sparsity.mask
+        templates_array = templates.get_dense_templates()
+    else:
+        templates_array = templates
```
Comment (samuelgarcia) [inline src/spikeinterface/curation/auto_merge.py:438]:
```suggestion
        templates_array = templates
        adaptative_masks = False
        sparsity = None
```

File: src/spikeinterface/curation/auto_merge.py [hunk 3]
```diff
@@ -403,30 +424,53 @@ def compute_templates_diff(sorting, templates, num_channels=5, num_shift=5, pair
...
-            template1 = templates[unit_ind1]
-            template2 = templates[unit_ind2]
+            template1 = templates_array[unit_ind1]
+            template2 = templates_array[unit_ind2]
             # take best channels
-            chan_inds = np.argsort(np.max(np.abs(template1 + template2), axis=0))[::-1][:num_channels]
+            if not adaptative_masks:
+                chan_inds = np.argsort(np.max(np.abs(template1 + template2), axis=0))[::-1][:num_channels]
+            else:
+                chan_inds = np.intersect1d(np.where(sparsity[unit_ind1])[0], np.where(sparsity[unit_ind2])[0])
```
Comment (samuelgarcia) [inline src/spikeinterface/curation/auto_merge.py:452]:
```suggestion
                chan_inds = np.intersect1d(np.flatnonzero(sparsity_mask[unit_ind1]), np.flatnonzero(sparsity_mask[unit_ind2]))
```

File: src/spikeinterface/curation/auto_merge.py [hunk 4]
```diff
@@ -489,3 +533,75 @@ def check_improve_contaminations_score(
...
+#                 sorting = sorting.select_units(to_keep)
+#                 break
+
+#             unit1, unit2 = merge
+#             sorting = MergeUnitsSorting(
+#                 sorting, [[unit1, unit2]], new_unit_ids=[unit1], delta_time_ms=censored_period_ms
+#             )
+#             subgraph = nx.contracted_nodes(subgraph, unit1, unit2, self_loops=False)
+#             subgraph.nodes[unit1]['score'] = highest_score
+#     return sorting
```
Comment (samuelgarcia) [inline src/spikeinterface/curation/auto_merge.py:607]:
```suggestion

```

**Comment by samuelgarcia** [comment]
@alejoe91 this is OK for me. Pierre will need this for circus.

**Comment by DradeAW** [comment]
I pulled the changed and ran the Lussac unit tests (which check that units that should be merged are indeed merged)

I get this error:
```
tests/modules/test_merge_units.py:51: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/lussac/modules/merge_units.py:56: in run
    potential_merges, extra_outputs = scur.get_potential_auto_merge(self.analyzer, extra_outputs=True, **params['auto_merge_params'])
../../dev/spikeinterface/spikeinterface/src/spikeinterface/curation/auto_merge.py:207: in get_potential_auto_merge
    templates = templates_ext.get_data(outputs="Templates")
../../dev/spikeinterface/spikeinterface/src/spikeinterface/core/sortinganalyzer.py:1704: in get_data
    return self._get_data(*args, **kwargs)
../../dev/spikeinterface/spikeinterface/src/spikeinterface/core/analyzer_extension_core.py:430: in _get_data
    probe=self.sorting_analyzer.get_probe(),
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = SortingAnalyzer: 16 channels - 15 units - 1 segments - memory - has recording
Loaded 2 extensions: random_spikes, templates

    def get_probe(self):
        probegroup = self.get_probegroup()
>       assert len(probegroup.probes) == 1, "There are several probes. Use `get_probegroup()`"
E       AssertionError: There are several probes. Use `get_probegroup()`

../../dev/spikeinterface/spikeinterface/src/spikeinterface/core/sortinganalyzer.py:776: AssertionError
```

**Comment by yger** [comment]
Aie. The bug seems to be more in the sorting_analyzer object. Or I should catch the fact that maybe the get_data() for the "templates" extension is not properly able to return Templates when more than one probes are selected. What do you think @samuelgarcia ? Otherwise, if more than one probe is detected, then we get Templates with the 'average' method, as it used to be done before. The following function will work properly since it can handled Templates or numpy arrays as input

**Comment by samuelgarcia** [comment]
@DradeAW I made a fix for multiple probe.
Can check this again please ?

**Comment by DradeAW** [comment]
All tests have passed on my end! :)

**Pull request merged**